packages/jest-parallel/src/worker.js

```diff
@@ -0,0 +1,166 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ * @format
+ */
+
+'use strict';
+
+const childProcess = require('child_process');
+
+import type {ChildProcess} from 'child_process';
+import type {Readable} from 'stream';
+
+import type {
+  JestChildMessage,
+  JestParentMessage,
+  JestQueueCallback,
+  JestQueueChildMessage,
+  JestWorkerOptions,
+} from './types';
+
+/**
+ * This class wraps the child process and provides a nice interface to
+ * communicate with. It takes care of:
+ *
+ *  - Re-spawning the process if it dies.
+ *  - Queues calls while the worker is busy.
+ *  - Re-sends the requests if the worker blew up.
+ *
+ * The reason for queueing them here (since childProcess.send also has an
+ * internal queue) is because the worker could be doing asynchronous work, and
+ * this would lead to the child process to read its receiving buffer and start a
+ * second call. By queueing calls here, we don't send the next call to the
+ * children until we receive the result of the previous one.
+ *
+ * As soon as a request starts to be processed by a worker, its "processed"
+ * field is changed to "true", so that other workers which might encounter the
+ * same call skip it.
+ */
+class JestWorker {
+  _busy: boolean;
+  _child: ChildProcess;
+  _options: JestWorkerOptions;
+  _queue: Array<JestQueueChildMessage>;
+
+  constructor(options: JestWorkerOptions) {
+    this._options = options;
+    this._queue = [];
+
+    this._initialize();
+  }
+
+  getStdout(): Readable {
+    return this._child.stdout;
+  }
+
+  getStderr(): Readable {
+    return this._child.stderr;
+  }
+
+  send(request: JestChildMessage, callback: JestQueueCallback) {
+    this._queue.push({callback, request});
+    this._process();
+  }
+
+  _initialize() {
+    const child = childProcess.fork(
+      require.resolve('./child'),
+      // $FlowFixMe: Flow does not work well with Object.assign.
+      Object.assign(
+        {
+          cwd: process.cwd(),
+          env: process.env,
+          silent: true,
+        },
+        this._options.forkOptions,
+      ),
+    );
+
+    child.on('message', this._receive.bind(this));
+    child.on('exit', this._exit.bind(this));
+
+    child.send({
+      file: this._options.workerPath,
+      type: 'initialize',
+    });
+
+    this._child = child;
+    this._busy = false;
```

Comment:
Does this needed to be here? AFAICT this could go into the constructor directly (if creating a process was asynchronous then it would make sense, but then you would also have to set `busy = true` at the beginning of the initialization and then call `_process()` at the end).

Reply:
The `initialize` method can be called multiple times; not only when instantiating `JestWorker`, but also if the child process unexpectedly dies. Once the process is restarted, it cannot be busy (whatever was the state of the previous process), so that's why it's set here.